### PR TITLE
fix(onnx): upgrade runtime for Spark 3.5 local inference

### LINKED ISSUE #2417: [BUG] ONNXModel fails in local Spark 3.5.6 environment
### SynapseML version

  1.0.13

### System information

  System Information:
  - Language version: Python 3.8+
  - Spark Version: 3.5.6
  - Spark Platform: Local Spark session (macOS)
  - OS: macOS (Darwin 24.6.0)

  Component(s) Affected:
  - ONNX Integration

  Language(s) Affected:
  - Python

### Describe the problem

  SynapseML's ONNXModel fails to execute ONNX models in local Spark environments with native library path errors, while
  the same code and models work successfully in Databricks environments. This suggests a configuration or environment
  setup issue specific to local Spark deployments.

  Expected Behavior:
  The ONNXModel.transform() method should work consistently across both local Spark environments and Databricks, given
  proper configuration.

  Actual Behavior:
  - ✅ Works in Databricks: Same ONNX models and SynapseML code execute successfully
  - ❌ Fails locally: Transformation fails with ONNX Runtime native library loading errors

### Code to reproduce issue

```python
  """
  Minimal Reproducible Example for SynapseML ONNX Issue

  Requirements:
  pip install pyspark synapseml scikit-learn onnx onnxruntime skl2onnx
  """

  import tempfile
  import os
  from sklearn.ensemble import RandomForestClassifier
  from sklearn.datasets import make_classification
  import numpy as np

  def create_simple_onnx_model():
      """Create a simple ONNX model for testing."""
      from skl2onnx import to_onnx
      from skl2onnx.common.data_types import FloatTensorType

      print("Creating simple scikit-learn model...")

      # Create simple dataset
      X, y = make_classification(
          n_samples=100, n_features=3, n_informative=2,
          n_redundant=1, n_classes=2, random_state=42
      )

      # Train simple model
      model = RandomForestClassifier(n_estimators=3, max_depth=3, random_state=42)
      model.fit(X, y)

      # Convert to ONNX
      initial_types = [('input', FloatTensorType([None, 3]))]
      onnx_model = to_onnx(model, X.astype(np.float32), initial_types=initial_types)

      # Save to temporary file
      temp_file = tempfile.NamedTemporaryFile(delete=False, suffix='.onnx')
      with open(temp_file.name, 'wb') as f:
          f.write(onnx_model.SerializeToString())

      print(f"ONNX model saved to: {temp_file.name}")
      return temp_file.name, X, y

  def test_synapseml_onnx():
      """Test ONNX model with SynapseML (fails locally, works in Databricks)."""
      from pyspark.sql import SparkSession
      from synapse.ml.onnx import ONNXModel
      from pyspark.ml.linalg import VectorUDT, Vectors
      from pyspark.sql.functions import udf

      print("\n=== Testing with SynapseML ONNXModel ===")

      # Create ONNX model
      onnx_path, X, y = create_simple_onnx_model()

      # Create Spark session (local configuration)
      print("Creating local Spark session...")
      spark = (SparkSession.builder
               .appName("SynapseML-ONNX-Bug-Repro")
               .config("spark.jars.packages", "com.microsoft.azure:synapseml_2.12:1.0.13")
               .config("spark.sql.adaptive.enabled", "false")
               .config("spark.driver.memory", "2g")
               .getOrCreate())

      spark.sparkContext.setLogLevel("WARN")

      try:
          # Convert test data to Spark DataFrame with Vector column
          def to_vector(row):
              return Vectors.dense(row)

          vector_udf = udf(to_vector, VectorUDT())

          # Create DataFrame
          test_data = X[:10].tolist()
          df = spark.createDataFrame(
              [(i, row) for i, row in enumerate(test_data)],
              ['id', 'features_list']
          )
          df = df.withColumn('features', vector_udf(df['features_list'])).select('id', 'features')

          # Configure SynapseML ONNX model
          onnx_ml = (ONNXModel()
                    .setModelLocation(onnx_path)
                    .setDeviceType("CPU")
                    .setFeedDict({"input": "features"})
                    .setFetchDict({"output_label": "prediction", "output_probability": "probability"}))

          # THIS IS WHERE IT FAILS IN LOCAL ENVIRONMENTS
          print("Attempting ONNX transformation...")
          result = onnx_ml.transform(df)

          print("✅ SynapseML ONNX transformation successful!")
          result.show(5)
          return True

      except Exception as e:
          print(f"❌ SynapseML ONNX transformation failed: {e}")
          print(f"Error type: {type(e).__name__}")
          return False

      finally:
          spark.stop()
          if os.path.exists(onnx_path):
              os.unlink(onnx_path)

  if __name__ == "__main__":
      test_synapseml_onnx()
```

### Other info / logs

  Additional Context:

  1. Reference Issue: This relates to the successful Databricks implementation documented in
  https://github.com/microsoft/SynapseML/issues/1902
  2. Local vs. Managed Environment: The core difference appears to be how ONNX Runtime native libraries are
  configured/discovered in local Spark vs. Databricks managed environments
  3. Configuration Attempts: Various local configuration attempts have been made:
    - Explicit java.library.path settings
    - Different ONNX Runtime Maven coordinates
    - Memory and execution tuning
    - Manual native library path specification
  4. Request for Documentation: It would be helpful to have documentation on the specific configuration requirements for
  local Spark environments to match Databricks functionality.

  Suggested Solution:

  Could SynapseML provide:
  1. Detailed local environment setup guide for ONNX integration
  2. Specific Maven/configuration requirements for local Spark deployments
  3. Native library path configuration examples for different platforms (macOS, Linux, Windows)

  This would help bridge the gap between managed Databricks environments and local development setups.

### What component(s) does this bug affect?

- [ ] `area/cognitive`: Cognitive project
- [x] `area/core`: Core project
- [x] `area/deep-learning`: DeepLearning project
- [ ] `area/lightgbm`: Lightgbm project
- [ ] `area/opencv`: Opencv project
- [ ] `area/vw`: VW project
- [ ] `area/website`: Website
- [ ] `area/build`: Project build system
- [ ] `area/notebooks`: Samples under notebooks folder
- [ ] `area/docker`: Docker usage
- [ ] `area/models`: models related issue

### What language(s) does this bug affect?

- [ ] `language/scala`: Scala source code
- [x] `language/python`: Pyspark APIs
- [ ] `language/r`: R APIs
- [ ] `language/csharp`: .NET APIs
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/synapse`: Azure Synapse integrations
- [ ] `integrations/azureml`: Azure ML integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by ranadeepsingh**
## Root cause

SynapseML's previous ONNX Runtime dependency was too old for the local Spark 3.5 inference scenario in #2417. ONNX Runtime's newer Java API also returns native-backed nested `OnnxValue` objects that require recursive materialization and deterministic cleanup.

## Change

- use cross-platform CPU `onnxruntime:1.17.3` by default, shipping Linux, Windows, macOS x64/ARM64, and Linux ARM64 natives
- keep GPU execution as an explicit `onnxruntime_gpu:1.17.3` opt-in and document mandatory exclusion of SynapseML's CPU artifact so exactly one ONNX Runtime jar is loaded
- normalize device selection and fail fast when explicit CUDA lacks a Spark GPU resource or a usable CUDA provider, rather than silently running on CPU
- recursively materialize nested ONNX sequence/map values and close child native handles safely
- close `SessionOptions` on success, fallback, and failure paths
- add dependency, lifecycle, CPU/CUDA selection, mixed-case, missing-resource, nested-value, and cleanup regression tests

## Validation

- real Databricks Tesla T4 validation with Spark-assigned executor GPU resource
- exactly one `onnxruntime_gpu:1.17.3` jar loaded; CPU, CUDA, and TensorRT providers available
- ORT profiling confirmed kernels executed through `CUDAExecutionProvider`
- CPU and CUDA predictions matched exactly
- T4 timings for the small validation model: CPU cold 1.9178 s / warm average 0.3000 s; CUDA cold 3.6622 s / warm average 0.2975 s
- 9/9 focused GH2417 tests passed; dependency/documentation guards and repeated native cleanup tests passed
- `deepLearning/codegen`, compile, and style checks passed
- full suite after `getDatasets`: modified branch 62 tests / 42 pass / 20 environmental failures; unmodified baseline 48 tests / 28 pass / the identical 20 failures
- GitHub compile, style, dependency review, CodeQL, and build checks passed on `8ce82aee39`

## Compatibility and risk

The default artifact is smaller and supports more operating systems than the GPU artifact. CPU and GPU ONNX Runtime artifacts expose the same Java classes and must not coexist; the documented exclusions enforce this. Explicit CUDA requests now fail clearly when requirements are unmet, while automatic selection retains fallback behavior.

Fixes #2417

### CODE REVIEW

File: deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala [hunk 1]
```diff
@@ -72,6 +72,13 @@ object ONNXUtils {
     }
   }
 
+  private def unwrapOnnxValue(value: OnnxValue): Any = {
+    value.getValue match {
+      case map: java.util.Map[_, _] => map.asScala.toMap
+      case other => other
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala]: Addressed in 49dce2103f713ed2399aa6efb5e3c2327f11e8f9. Sequence conversion now recursively materializes nested lists/maps before cleanup and closes every level of child OnnxValue handles in finally blocks. ONNXValueConverterSuite verifies nested values remain usable, direct/deep child handles close, and the outer result-owned value is left for OrtSession.Result.

File: deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXValueConverterSuite.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    val converted = ONNXValueConverter.mapSequenceToArray(outerSequence)
+
+    assert(converted == Vector(Vector(42L)))
+    assert(nestedSequence.closed)
+    assert(leaf.closed)
+    assert(!outerSequence.closed)
+  }
+
+  private object StubValueInfo extends ValueInfo
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXValueConverterSuite.scala:28]: Verified against the resolved ONNX Runtime 1.16.3 API: ai.onnxruntime.ValueInfo is a marker interface with no methods; getType() belongs to OnnxValue, and StubOnnxValue implements it below. Both deepLearning/Test/compile and ONNXValueConverterSuite pass against onnxruntime_gpu:1.16.3, so no code change is needed for this comment.

File: project/OnnxRuntimeDependency.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  *
+  * GH2417: 1.8.1 fails to load its natives for local Spark 3.5 inference, and the GPU-only artifact never
+  * ships macOS natives at all (CUDA has no macOS support), so upgrading `onnxruntime_gpu` alone cannot fix
+  * macOS. 1.17.3 is confirmed to fix local CPU inference and additionally publishes a CPU-only `onnxruntime`
+  * artifact with macOS x64/aarch64 natives, so it is used as the default cross-platform dependency in
+  * build.sbt. See docs/Explore Algorithms/Deep Learning/ONNX.md for the CUDA opt-in instructions.
+  */
+object OnnxRuntimeDependency {
+  val Version = "1.17.3"
+  private val Organization = "com.microsoft.onnxruntime"
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply project/OnnxRuntimeDependency.scala:28]: Resolved: the PR description now matches the shipped implementation—cross-platform CPU onnxruntime:1.17.3 by default, explicit onnxruntime_gpu:1.17.3 opt-in with exclusions, and the completed T4/packaging/lifecycle validation.
